File: e28/java/Solution.java
public class Solution {
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        // write your code here
        if (matrix == null || matrix.length == 0) {
            return false;
        }
        int m = matrix.length;
        int n = matrix[0].length;
        if (target < matrix[0][0] || target > matrix[m-1][n-1]) {
            return false;
        }
        //locate line
        int left = 0;
        int right = m - 1;
        int possibleLine = 0;
        if (target > matrix[m-1][0]) {
            possibleLine = m - 1;
        }
        else {
            while (right - left > 1) {
                int mid = (left + right) / 2;
                if (matrix[mid][0] <= target) {
                    left = mid;
                }
                else {
                    right = mid;
                }
            }
            possibleLine = left;
        }
        //locate column
        if (target == matrix[possibleLine][0] || target ==                                matrix[possibleLine][n-1]) {
            return true;
        }
        int l = 0;
        int r = n - 1;
        while (r - l > 1) {
            int midd = (l + r) / 2;
            if (matrix[possibleLine][midd] == target) {
                return true;
            }
            else if (matrix[possibleLine][midd] < target) {
                l = midd;
            }
            else {
                r = midd;
            }
        }
        return false;
    }
}

File: e28/python/solution.py
class Solution:
    """
    @param matrix, a list of lists of integers
    @param target, an integer
    @return a boolean, indicate whether matrix contains target
    """
    def searchMatrix(self, matrix, target):
        # write your code here
        if matrix is None or matrix == []:
            return False
        m = len(matrix)
        n = len(matrix[0])
        #locate line
        firstColumn = [matrix[line][0] for line in range(m)]
        if target < firstColumn[0]:
            return False
        if target > firstColumn[-1]:
            possibleLine = matrix[-1]
        else:
            left = 0
            right = m - 1
            while right - left > 1:
                mid = (left + right) / 2
                if firstColumn[mid] == target:
                    return True
                elif firstColumn[mid] < target:
                    left = mid
                else:
                    right = mid
            possibleLine = matrix[left]
        
        #locate column
        l = 0
        r = n - 1

        if target < possibleLine[0] or target > possibleLine[-1]:
            return False
        if target == possibleLine[0] or target == possibleLine[-1]:
            return True
        while r - l > 1:
            mid = (l + r) / 2
            if possibleLine[mid] == target:
                return True
            elif possibleLine[mid] < target:
                l = mid
            else:
                r = mid
        return False
